perceval/utils/noise_model.py

```diff
@@ -0,0 +1,109 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from ._validated_fields import AValidatedParam, ValidatedBool, ValidatedFloat
+from typing import Dict
+
+
+class NoiseModel:
+    """
+    The NoiseModel class contains all noise parameters which can be supported by Perceval. Default value of each
+    parameter means "no noise", so a NoiseModel constructed with all default parameters leads to a perfect simulation.
+
+    :param brightness: first lens brightness of a quantum dot
+    :param indistinguishability: chance two photons are indistinguishable
+    :param g2: chance two photons are emitted instead of one (g²(0) value)
+    :param g2_distinguishable: g2-generated photons indistinguishability
+    :param transmittance: system-wide transmittance (warning, can interfere with the brightness parameter)
+    :param phase_imprecision: maximum precision of the phase shifters (0 means infinite precision)
+    """
+
+    def __init__(self,
+                 brightness: float = None,
+                 indistinguishability: float = None,
+                 g2: float = None,
+                 g2_distinguishable: bool = None,
+                 transmittance: float = None,
+                 phase_imprecision: float = None
+                 ):
+        # Source parameters
+        self._params: Dict[str, AValidatedParam] = {}
+        self._add_param(ValidatedFloat("brightness", brightness, 0, 1, 1))
+        self._add_param(ValidatedFloat("indistinguishability", indistinguishability, 0, 1, 1))
+        self._add_param(ValidatedFloat("g2", g2, 0, 1, 0))
+        self._add_param(ValidatedBool("g2_distinguishable", g2_distinguishable, True))
+
+        # System-wide parameter
+        self._add_param(ValidatedFloat("transmittance", transmittance, 0, 1, 1))
+
+        # Optical circuit parameter
+        self._add_param(ValidatedFloat("phase_imprecision", phase_imprecision, 0, default_value=0))
+
+    def _add_param(self, param: AValidatedParam):
+        self._params[param.name] = param
+        cls = type(self)
+        if not hasattr(cls, param.name):
+            setattr(cls, param.name, property(lambda slf: slf._params[param.name].get()))
+
+    def __getitem__(self, param_name: str):
+        if param_name in self._params:
+            return self._params[param_name]
+        raise KeyError(f"No parameter named '{param_name}'")
+
+    def set_value(self, param_name: str, value):
+        self[param_name].set(value)
+
+    def __str__(self):
+        return str(self.__dict__())
+
+    def __dict__(self):
+        return {k: v.get() for k, v in self._params.items() if not v.is_default}
+
+    def __eq__(self, other):
+        for field in self._params:
+            if self._params[field] != other._params[field]:
+                return False
+        return True
+
+
+if __name__ == "__main__":
```

Comment:
You have this segment to debug while developing by running as a standalone script, isn't it?

When I tried running the `noise_model.py` as a standalone script, it did not work.

Reply:
You're right, I forgot to clean up my file before committing.